sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,436 @@
+
+  // (Required) The CombineFn for this Combine
+  FunctionSpec combine_fn = 1;
+
+  // (Required) A reference to the Coder to use for accumulators of the CombineFn
+  int64 accumulator_coder_ref = 2;
+
+  // (Optional) if the CombineFn uses side inputs, a map from the id to the spec
+  // for how to materialize and read the PCollection as a side input
+  map<string, SideInput> side_inputs = 3;
+}
+
+// A coder, the binary format for serialization and deserialization of data in
+// a pipeline.
+message Coder {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A cross-language, stable, unique identifier for the (possibly
+  // parametric) encoding.
+  string urn = 2;
+
+  // (Optional) If this coder is parametric, such as ListCoder(VarIntCoder),
+  // this is a list of the components. In order for encodings to be identical,
+  // the URN and all components must be identical, recursively.
+  repeated int64 component_coder_reference = 3;
+
+  // (Optional) An SDK-specific UDF implementing the coder.
+  //
+  // TODO: reconsider how to do this, aka CustomCoder
+  FunctionSpec custom_coder_fn = 4;
+}
+
+// A windowing strategy describes the window function, triggering, allowed
+// lateness, and accumulation mode for a PCollection.
+//
+// TODO: consider inlining field on PCollection if they aren't really a coherent whole
+message WindowingStrategy {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) The UDF assigns windows, merges windows, and shifts timestamps
+  // before they are combined according to the OutputTime.
+  FunctionSpec window_fn = 2;
+
+  // (Required) Whether or not the window fn is merging.
+  //
+  // This knowledge is required for many optimizations.
+  bool is_merging_window_fn = 3;
+
+  // (Required) The coder for the windows of this PCollection.
+  int64 window_coder_reference = 4;
+
+  // (Required) The trigger, serialized as a simple syntax tree.
+  //
+  // TODO: say more
+  Trigger trigger = 5;
+
+  // (Required) The accumulation mode indicates whether new panes are a full
+  // replacement for prior panes or whether they are deltas to be combined
+  // with other panes (the combine should correspond to whatever the upstream
+  // grouping transform is).
+  AccumulationMode accumulation_mode = 6;
+
+  // (Required) The OutputTime specifies, for a grouping transform, how to
+  // compute the aggregate timestamp. The window_fn will first possibly shift
+  // it later, then the OutputTime takes the max, min, or ignores it and takes
+  // the end of window.
+  //
+  // This is actually only for input to grouping transforms, but since they
+  // may be introduced in runner-specific ways, it is carried along with the
+  // windowing strategy.
+  OutputTime output_time = 7;
+
+  // (Required) The duration beyond the end of a window at which the window becomes droppable.
+  int64 allowed_lateness = 8;
+}
+
+enum AccumulationMode {
```

Comment:
I think these enums could  be defined within WindowingStrategy since they're only useful there?